Source/FieldSolver/ImplicitSolvers/WarpXImplicitOps.cpp

```diff
@@ -0,0 +1,333 @@
+
+#include <AMReX.H>
+#include <AMReX_Array.H>
+#include <AMReX_BLassert.H>
+#include <AMReX_Geometry.H>
+#include <AMReX_IntVect.H>
+#include <AMReX_LayoutData.H>
+#include <AMReX_MultiFab.H>
+#include <AMReX_ParmParse.H>
+#include <AMReX_Print.H>
+#include <AMReX_REAL.H>
+#include <AMReX_Utility.H>
+#include <AMReX_Vector.H>
+
+#include <algorithm>
+#include <array>
+#include <memory>
+#include <ostream>
+#include <vector>
+
+void
+WarpX::ImplicitPreRHSOp ( amrex::Real  a_cur_time,
+                          amrex::Real  a_full_dt,
+                          int          a_nl_iter,
+                          bool         a_from_jacobian )
+{
+    using namespace amrex::literals;
+    amrex::ignore_unused( a_full_dt, a_nl_iter, a_from_jacobian );
+
+    // Advance the particle positions by 1/2 dt,
+    // particle velocities by dt, then take average of old and new v,
+    // deposit currents, giving J at n+1/2
+    // This uses Efield_fp and Bfield_fp, the field at n+1/2 from the previous iteration.
+    const bool skip_current = false;
+    const PushType push_type = PushType::Implicit;
+    PushParticlesandDeposit(a_cur_time, skip_current, push_type);
+
+    SyncCurrentAndRho();
+
+}
+
+void
+WarpX::SetElectricFieldAndApplyBCs ( const WarpXSolverVec&  a_E )
+{
+    const amrex::Vector<std::array< std::unique_ptr<amrex::MultiFab>, 3 > >& Evec = a_E.getVec();
+    amrex::MultiFab::Copy(*Efield_fp[0][0], *Evec[0][0], 0, 0, ncomps, Evec[0][0]->nGrowVect());
+    amrex::MultiFab::Copy(*Efield_fp[0][1], *Evec[0][1], 0, 0, ncomps, Evec[0][1]->nGrowVect());
+    amrex::MultiFab::Copy(*Efield_fp[0][2], *Evec[0][2], 0, 0, ncomps, Evec[0][2]->nGrowVect());
+    FillBoundaryE(guard_cells.ng_alloc_EB, WarpX::sync_nodal_points);
+    ApplyEfieldBoundary(0, PatchType::fine);
+}
+
+void
+WarpX::UpdateMagneticFieldAndApplyBCs( const amrex::Vector<std::array< std::unique_ptr<amrex::MultiFab>, 3 > >&  a_Bn,
+                            amrex::Real                                                         a_thetadt )
+{
+    amrex::MultiFab::Copy(*Bfield_fp[0][0], *a_Bn[0][0], 0, 0, ncomps, a_Bn[0][0]->nGrowVect());
+    amrex::MultiFab::Copy(*Bfield_fp[0][1], *a_Bn[0][1], 0, 0, ncomps, a_Bn[0][1]->nGrowVect());
+    amrex::MultiFab::Copy(*Bfield_fp[0][2], *a_Bn[0][2], 0, 0, ncomps, a_Bn[0][2]->nGrowVect());
+    EvolveB(a_thetadt, DtType::Full);
+    ApplyMagneticFieldBCs();
+}
+
+void
+WarpX::FinishMagneticFieldAndApplyBCs( const amrex::Vector<std::array< std::unique_ptr<amrex::MultiFab>, 3 > >&  a_Bn,
+                            amrex::Real                                                         a_theta )
+{
+    FinishImplicitField(Bfield_fp, a_Bn, a_theta);
+    ApplyMagneticFieldBCs();
+}
+
+void
+WarpX::ApplyMagneticFieldBCs()
+{
+    FillBoundaryB(guard_cells.ng_alloc_EB, WarpX::sync_nodal_points);
+    ApplyBfieldBoundary(0, PatchType::fine, DtType::Full);
+}
+
+void
+WarpX::SaveParticlesAtImplicitStepStart ( )
+{
```

Comment:
Could you add a few lines of comments explaining what this does?
(i.e. that, in the particle iterations required for the particle-suppressed schemes, we need to iterate to find the position and velocity at the half-time step, and this iteration requires knowledge of the particles position and momentum at the beginning of the time step. Thus, we need to save this information.)